src/hdf5_dl.c

```diff
@@ -97,44 +97,167 @@ hid_t H5E_ERR_CLS_g = -1;
+            if(symbol != NULL)
+                DL_H5Functions.H5Pget_filter_by_id2 = (DL_func_H5Pget_filter_by_id2)symbol;
+            else
+              retval = 0;
+        }
+
+        if(DL_H5Functions.H5Pget_chunk == NULL) {
+            symbol = find_sym(handle, "H5Pget_chunk");
+            if(symbol != NULL)
+                DL_H5Functions.H5Pget_chunk = (DL_func_H5Pget_chunk)symbol;
+            else
+              retval = 0;
+        }
+
+        if(DL_H5Functions.H5Pmodify_filter == NULL) {
+            symbol = find_sym(handle, "H5Pmodify_filter");
+            if(symbol != NULL)
+                DL_H5Functions.H5Pmodify_filter = (DL_func_H5Pmodify_filter)symbol;
+            else
+              retval = 0;
+        }
+
         /*H5T*/
-        DL_H5Functions.H5Tget_size = (DL_func_H5Tget_size)dlsym(handle, "H5Tget_size");
-        DL_H5Functions.H5Tget_class = (DL_func_H5Tget_class)dlsym(handle, "H5Tget_class");
-        DL_H5Functions.H5Tget_super = (DL_func_H5Tget_super)dlsym(handle, "H5Tget_super");
-        DL_H5Functions.H5Tclose = (DL_func_H5Tclose)dlsym(handle, "H5Tclose");
+        if(DL_H5Functions.H5Tget_size == NULL) {
+            symbol = find_sym(handle, "H5Tget_size");
+            if(symbol != NULL)
+                DL_H5Functions.H5Tget_size = (DL_func_H5Tget_size)symbol;
+            else
+              retval = 0;
+        }
+
+        if(DL_H5Functions.H5Tget_class == NULL) {
+            symbol = find_sym(handle, "H5Tget_class");
+            if(symbol != NULL)
+                DL_H5Functions.H5Tget_class = (DL_func_H5Tget_class)symbol;
+            else
+              retval = 0;
+        }
+       
+        if(DL_H5Functions.H5Tget_super == NULL) {
+            symbol = find_sym(handle, "H5Tget_super");
+            if(symbol != NULL)
+                DL_H5Functions.H5Tget_super = (DL_func_H5Tget_super)symbol;
+            else
+              retval = 0;
+        }
+        
+        if(DL_H5Functions.H5Tclose == NULL) {
+            symbol = find_sym(handle, "H5Tclose");
+            if(symbol != NULL)
+                DL_H5Functions.H5Tclose = (DL_func_H5Tclose)symbol;
+            else
+              retval = 0;
+        }
+
         /*H5Z*/
-        DL_H5Functions.H5Zregister = (DL_func_H5Zregister)dlsym(handle, "H5Zregister");
+        if(DL_H5Functions.H5Zregister == NULL) {
+            symbol = find_sym(handle, "H5Zregister");
+            if(symbol != NULL)
+                DL_H5Functions.H5Zregister = (DL_func_H5Zregister)symbol;
+            else
+              retval = 0;
+        }
 
         /*Variables*/
-        H5E_CANTREGISTER_g = *((hid_t *)dlsym(handle, "H5E_CANTREGISTER_g"));
-        H5E_CALLBACK_g = *((hid_t *)dlsym(handle, "H5E_CALLBACK_g"));
-        H5E_PLINE_g = *((hid_t *)dlsym(handle, "H5E_PLINE_g"));
-        H5E_ERR_CLS_g = *((hid_t *)dlsym(handle, "H5E_ERR_CLS_g"));
-
-        is_init = true;
+        if(H5E_CANTREGISTER_g == -1) {
```

Comment:
You're assuming the HDF5 devs won't ever assign the value -1 to this symbol (or any of the other `..._g` variables) in their library, which I don't think you can do.  There's no value of this variable which you can use as a sigil to indicate it hasn't been resolved because there are no values available to you that aren't available to them.  You need something else out-of-band to track whether the value is properly initialised or not.

Reply:
The easiest might be to just collect all the pointers for these variables (the ones return by dlsym), which you _can_ distinguish from NULL, and then just dereference them all at the end once you've determined you've found everything.

Reply:
@ketiltrout, I've added pointers to address this issue now